research/clustering/information_abstraction.py

```diff
@@ -0,0 +1,439 @@
+        :param num_simulations: how many simulations you want to do
+        :return: [win_rate, loss_rate, tie_rate]
+        """
+        ehs = [0] * 3
+        for _ in range(num_simulations):
+
+            idx = game.get_winner()
+
+            # increment win rate for winner/tie
+            ehs[idx] += 1 / num_simulations
+
+        return ehs
+
+    def simulate_get_turn_ehs_distributions(self, available_cards: List[int], the_board: List[int],
+                                            our_hand: List[int], num_simulations: int = 5) -> np.array:
+        """
+        # TODO num_simulations should be higher
+
+        :param available_cards: list of available cards on the turn
+        :param the_board: the board as of the turn
+        :param our_hand: cards our hand (Card.eval_card)
+        :param num_simulations: int of simulations
+        :return: array of counts for each cluster the turn fell into by the river after simulations
+        """
+        turn_ehs_distribution = [0] * len(self._river_centroids)
+
+        # sample river cards and run a simulation
+        for _ in range(num_simulations):
+
+            river_card = random.sample(available_cards, 1)
+            board = list(the_board)  # copy list
+            board = board + river_card
+
+            game = GameUtility(
+                our_hand=our_hand, board=board, cards=self._evals
+            )
+            ehs = self.simulate_get_ehs(game)
+
+            # get EMD for expected hand strength against each river centroid
+            # to which does it belong?
+            for idx, river_centroid in enumerate(self._river_centroids):
+                # TODO: do we need a faster implementation of this?
+                emd = wasserstein_distance(ehs, river_centroid)
+
+                if idx == 0:
+                    min_idx = idx
+                    min_emd = emd
+                else:
+                    if emd < min_emd:
+                        min_idx = idx
+                        min_emd = emd
+
+            # now increment the cluster to which it belongs -
+            turn_ehs_distribution[min_idx] += (
+                    1 / num_simulations
+            )
+        return turn_ehs_distribution
+
+    def get_river_ehs(self, num_print: int) -> np.ndarray:
+        """
+
+        :param num_print: number of simulations of opponents cards for calculating ehs
+        :return: np.ndarray of arrays containing [win_rate, loss_rate, tie_rate]
+        """
+        start = time.time()
+        river_ehs = [0] * len(self.river)
+
+        # iterate over possible boards/hole cards
+        for i, public in enumerate(self.river):
+
+            our_hand = list(public[:2])
+            board = list(public[2:7])
+
+            # get expected hand strength
+            game = GameUtility(our_hand=our_hand, board=board, cards=self._evals)
+            river_ehs[i] = self.simulate_get_ehs(game)
+
+            if i % num_print == 0:
+                print(
```

Comment:
this would need to be `tqdm.write` or deal with sys.stdout (see bottom of tqdm README)

Reply:
sounds good